lib/cli/src/Cardano/CLI.hs

```diff
@@ -296,6 +331,202 @@ runCli = join . customExecParser preferences
+-- another type, to include Shelley keys.
+parseKeyType :: Parser (IO ByronWalletStyle)
+parseKeyType = toIO . fromText <$> strOption
+    ( long "type"
+    <> metavar "KEYTYPE"
+    <> helpDoc (Just (vsep typeOptions))
+    )
+  where
+    toIO :: Either TextDecodingError ByronWalletStyle -> IO ByronWalletStyle
+    toIO (Right x) =
+        return x
+    toIO (Left (TextDecodingError e)) =
+        hPutStrLn stderr e >> exitFailure
+
+    typeOptions = string <$>
+        [ "Any of the following:"
+        , "  random (" ++ allowedWords Random ++ ")"
+        , "  icarus (" ++ allowedWords Icarus ++ ")"
+        , "  trezor (" ++ allowedWords Trezor ++ ")"
+        , "  ledger (" ++ allowedWords Ledger ++ ")"
+        ]
+
+    allowedWords = (++ " words")
+        . formatEnglishEnumeration
+        . map show
+        . allowedWordLengths
+        . byronCliKeyScheme
+       where
+          -- >>> formatEnglishEnumeration ["a", "b", "c"]
+          -- "a, b or c"
+          --
+          -- >>> formatEnglishEnumeration ["a", "b"]
+          -- "a or b"
+          --
+          -- >>> formatEnglishEnumeration ["a"]
+          -- "a"
+         formatEnglishEnumeration = formatEnglishEnumerationRev . reverse
+         formatEnglishEnumerationRev [ult, penult]
+            = penult ++ " or " ++ ult
+         formatEnglishEnumerationRev (ult:penult:revBeginning)
+            = intercalate ", " (reverse revBeginning)
+                ++ ", "
+                ++ penult
+                ++ " or "
+                ++ ult
+         formatEnglishEnumerationRev xs = intercalate ", " (reverse xs)
+
+parseSeedWords :: Parser [Text]
+parseSeedWords = some (argument str (metavar "MNEMONIC_WORDS..."))
+
+-- | Record with mnemonic and key derivation funcionality — /without/ any type
+-- parameters related to scheme.
+--
+-- This means that we can have a value for byron, a value for icarus, both with
+-- the same type @CliKeyScheme@.
+--
+-- @CliKeyScheme@ is on the other hand parameterized over @key@ and @m@.
+-- @hoistKeyScheme@ is provided for mapping over @m@. @mapKey@ is provided for
+-- mapping over @key@.
+--
+-- This way we can test mapping @XPrv@s to @Text@ as pure code in
+-- @Either String@, rather than @IO@.
+data CliKeyScheme key m = CliKeyScheme
+    { allowedWordLengths :: [Int]
+    , mnemonicToRootKey :: [Text] -> m key
+    }
+
+-- | Change the underlying monad of a @CliKeyScheme@.
+hoistKeyScheme
+    :: (forall a. m1 a -> m2 a)
+    -> CliKeyScheme key m1
+    -> CliKeyScheme key m2
+hoistKeyScheme eta s = CliKeyScheme
+    { allowedWordLengths = allowedWordLengths s
+    , mnemonicToRootKey = eta . mnemonicToRootKey s
+    }
+
+xPrvToHexText :: XPrv -> Text
+xPrvToHexText = T.pack . B8.unpack . hex . unXPrvStripPub
+
+unsafeHexTextToXPrv :: Text -> XPrv
+unsafeHexTextToXPrv t = either error Prelude.id $
+    xPrvFromStrippedPubXPrv
+    . unsafeFromHex
+    . B8.pack
+    . T.unpack
+    $ t
```

Comment:
What about having that in `Cardano.Wallet.Unsafe` ?

Reply:
I'm inclined to keep it here because:
- I want it next to `xPrvToHexText` as a dual.
- I had thoughts about making `mapKey` take a bidirectional map that can fail* `(key1 -> m key2, key2 -> m key1)`. It would not be `unsafe` in that case. (* more on that later)